### ISSUE TITLE
Collision detection improvements

### ISSUE THREAD

**Issue by st-pasha** [flame-engine/flame#1835]
This is a proposal for some changes to the hitbox/collision-detection system, for the purpose of improving efficiency and interoperability with other parts of Flame. This issue is intended as discussion / collaboration with @spydon, before we start the actual work.

First problem that needs to be addressed is that currently the collision detection system can only "live" at the root FlameGame level, which is not always optimal. In a more complicated game the game world can be several levels deep in the component tree, or there could be several worlds (say, for each level). So, I suggest then to 

- [ ] implement a mixin similar to `HasCollisionDetection`, but which could be applied to any `Component`.
    - This could (in theory) be as easy as changing the `on` part of the current mixin.

Second problem is that the current implementation of `ShapeHitbox` is based on `ShapeComponent`s, not on `Shape`s. Thus, we need to

- [ ] create `Hitbox`es based on `Shape` class and its descendants.
    - The API of both classes is quite similar already. 
    - This would allow us to test for intersections more efficiently (the `Shape` class has support for intersection detection of convex shapes).
        - In order to take advantage of this algorithm, we'd need to change the API of the current `intersections()` method; or, rather, create a second more efficient method and deprecate the first. The signature of the new method would be `Intersection? intersection(T other);`, which is to say it returns `null` if the two shapes do not intersection, or an `Intersection` object if they do. The `Intersection` object can then be queried for the intersection points, or some other intersection properties. However, if the user is not interested in the intersection points - then that calculation can be skipped.
            - the `Intersection` class could also contain references to the hitboxes + components that are colliding.
            - note that currently `Hitbox` class is defined as `implements GenericCollisionsCallbacks<T>`, where `T` is the derived hitbox type. However, the `GenericCollisionsCallbacks` class says that `T` has to be the type of the object to which the hitbox is attached. It looks like a bug to me.
        - the `possiblyIntersects()` method can be deprecated as well, since the `intersection()` method covers that use-case. 

The biggest source of inefficiency with the current collision detection system is that whenever any two hitboxes need to be checked for intersection, they are projected into the "absolute" coordinates (i.e. the coordinate system of the canvas). This is expensive: not only some objects never move and therefore don't have to be projected each time, if there are `N` objects, we currently end up doing `O(N^2)` projections, whereas simply performing a projection once at the start of a tick and then computing the intersections on those would only be `O(N)`.

- [ ] Cache projections of each hitbox onto the common collision plane.
    - The "collision plane" here is the common ancestor of all hitboxes, which is marked with the new `HasCollisionDetection` mixin (see item 1 above).
    - Projections are easy for hitboxes based on `Shape` class, which already implements the `project()` method.
    - We would need to keep track of which components have moved in the last tick, and re-calculate their projections.

Current `ShapeComponent` class (and descendants) contains both the "shape-component" logic, and also the logic for computing intersections between various shapes. This is too jumbled: we should

- [ ] Implement clean `ShapeComponent` class to be used in examples etc.
    - Only deals with drawing / moving a shape;
    - Supports multiple paints (so that you can have a rectangle with fill, border and shadow);
    - Has a simple flag to also add a hitbox, but otherwise handles hitboxes in the same way as other components do.

**Comment by spydon** [comment]
>     * [ ]  implement a mixin similar to `HasCollisionDetection`, but which could be applied to any `Component`.

Sounds good, should be fairly simple to implement.

> Second problem is that the current implementation of `ShapeHitbox` is based on `ShapeComponent`s, not on `Shape`s. Thus, we need to
> 
>     * [ ]  create `Hitbox`es based on `Shape` class and its descendants.

The problem with this is that we want to be able to add hitboxes as components and we want the shape class to be more generic and possible to be used outside of FCS. Otherwise we'll go back to how the first collision detection system was, with `addHitbox` and I think the current system is nicer for the user.
     
>           * the `Intersection` class could also contain references to the hitboxes + components that are colliding.

We'll have to make sure that this class is very cheap to instantiate then, since we'll be creating a lot of these objects.

>           * note that currently `Hitbox` class is defined as `implements GenericCollisionsCallbacks<T>`, where `T` is the derived hitbox type. However, the `GenericCollisionsCallbacks` class says that `T` has to be the type of the object to which the hitbox is attached. It looks like a bug to me.

I haven't looked into the code for this for quite some time, but how would it possibly work currently if they wouldn't end up with matching types?

> The biggest source of inefficiency with the current collision detection system is that whenever any two hitboxes need to be checked for intersection, they are projected into the "absolute" coordinates (i.e. the coordinate system of the canvas). This is expensive: not only some objects never move and therefore don't have to be projected each time, if there are `N` objects, we currently end up doing `O(N^2)` projections, whereas simply performing a projection once at the start of a tick and then computing the intersections on those would only be `O(N)`.
> 
>     * [ ]  Cache projections of each hitbox onto the common collision plane.

I agree it is ineffiecient, the projections should be cached currently too though.

>       * Projections are easy for hitboxes based on `Shape` class, which already implements the `project()` method.

This is the big question here I think, and this is what we should start with before we start with this work. Since we currently don't have an efficient way of getting the current `Transform2D` this will be quite inefficient too. I think that we have to come up with a way to always track this similarly to #1684 but more efficiently, which will also lead to that we can set global render priorities.

> Current `ShapeComponent` class (and descendants) contains both the "shape-component" logic, and also the logic for computing intersections between various shapes. This is too jumbled: we should

Agree, it grew organically into what it is today.

**Comment by st-pasha** [comment]
> The problem with this is that we want to be able to add hitboxes as components and we want the shape class to be more generic and possible to be used outside of FCS. Otherwise we'll go back to how the first collision detection system was, with addHitbox and I think the current system is nicer for the user.

Yeah, we may need to think this through a bit more carefully. Let's stick with the idea that from a user's perspective a "hitbox" is a component: call it `HitboxComponent`. This component contains: a `Shape` of the hitbox, and the `collisionType` property. Now, the collision detection system keeps a reference to this object, and also creates projections of the hitbox shape onto the root plane. We could probably keep the projection inside the hitbox as well. So, the resulting structure would be either
```dart
class HitboxComponent extends Component {
  ShapeHitbox hitbox;
}
class ShapeHitbox implements Hitbox<> {
  Component owner;
  CollisionType collisionType;
  Shape shape;
  Shape globalShape;
}
```
or
```dart
class HitboxComponent extends Component implements Hitbox<> {
  CollisionType collisionType;
  Shape shape;
  Shape globalShape;
}
```
I'm not yet sure which one is better.


> We'll have to make sure that this class is very cheap to instantiate then, since we'll be creating a lot of these objects.

It should be cheaper than the `Set<>` object that we're currently creating.

> I haven't looked into the code for this for quite some time, but how would it possibly work currently if they wouldn't end up with matching types?

My understanding is that one `T` is a `PositionComponent`, while the other is a `ShapeHitbox` which is a mixin on `ShapeComponent` derived from `PositionComponent` -- so it kinda works, but it wouldn't generalize correctly to non-component-based hitboxes, I think.

> the projections should be cached currently too though.

Looks like they're only partially cached. For example, `PolygonHitbox` which is based on `PolygonComponent` checks the cache validity based on its `absolutePosition`, `absoluteAngle` and `absoluteScale` -- and these properties aren't cached. So we end up recalculating them for every intersection check, even though we don't recalculate the entire polygon coordinates.

**Comment by ufrshubham** [comment]
Sometime back there was a discussion about introducing a concept of layers for hitboxes, but it was shot down at that time 😅. I'm hoping that it gets considered in one of these improvements some day. Some use cases do not work very well with just the existing 3 `CollisionType`s and having collision layers would benefit such cases. #980
